src/software/ai/navigator/navigator.cpp

```diff
@@ -2,303 +2,285 @@
-        this->world.friendlyTeam().getRobotById(move_intent.getRobotId())->position();
-    Point dest = move_intent.getDestination();
-
-    std::vector<Obstacle> obstacles =
-        createCurrentObstacles(move_intent.getAreasToAvoid(), move_intent.getRobotId());
-
-    Rectangle navigable_area(
-        Point(this->world.field().totalXLength(), this->world.field().totalYLength()),
-        this->world.field().totalXLength(), this->world.field().totalYLength());
-
-    Path path = path_planner->findPath(start, dest, navigable_area, obstacles);
-
-    moveNavigation(move_intent, path);
+    move_intents.push_back(intent);
+    current_primitive = std::unique_ptr<Primitive>(nullptr);
 }
 
-void Navigator::visit(const MoveSpinIntent &move_spin_intent)
+void Navigator::visit(const MoveSpinIntent &intent)
 {
-    auto p            = std::make_unique<MoveSpinPrimitive>(move_spin_intent);
+    auto p            = std::make_unique<MoveSpinPrimitive>(intent);
     current_primitive = std::move(p);
+    registerNonMoveIntentRobotId(intent.getRobotId());
 }
 
-void Navigator::visit(const PivotIntent &pivot_intent)
+void Navigator::visit(const PivotIntent &intent)
 {
-    auto p            = std::make_unique<PivotPrimitive>(pivot_intent);
+    auto p            = std::make_unique<PivotPrimitive>(intent);
     current_primitive = std::move(p);
+    registerNonMoveIntentRobotId(intent.getRobotId());
 }
 
-void Navigator::visit(const StopIntent &stop_intent)
+void Navigator::visit(const StopIntent &intent)
 {
-    auto p            = std::make_unique<StopPrimitive>(stop_intent);
+    auto p            = std::make_unique<StopPrimitive>(intent);
     current_primitive = std::move(p);
+    registerNonMoveIntentRobotId(intent.getRobotId());
 }
 
-// helpers
-void Navigator::moveNavigation(const MoveIntent &move_intent, const Path &path)
+std::vector<std::unique_ptr<Primitive>> Navigator::getAssignedPrimitives(
+    const World &world, const std::vector<std::unique_ptr<Intent>> &assignedIntents)
 {
-    if (path)
+    this->world = world;
+    planned_paths.clear();
+    move_intents.clear();
+    friendly_non_move_intent_robot_obstacles.clear();
+
+    auto assigned_primitives = std::vector<std::unique_ptr<Primitive>>();
+    for (const auto &intent : assignedIntents)
     {
-        std::vector<Point> path_points = path->getKnots();
-        planned_paths.emplace_back(path_points);
-        if (path_points.size() == 1)
+        intent->accept(*this);
+        if (current_primitive)
         {
-            throw std::runtime_error(
-                "Path only contains one point, which is invalid, since it's ambiguous if it's the start or dest or some other point");
+            assigned_primitives.emplace_back(std::move(current_primitive));
         }
-        if (path_points.size() == 2)
+    }
+
+    addMoveIntentsToAssignedPrimitives(assigned_primitives);
+
+    return assigned_primitives;
+}
+
+std::set<PathObjective> Navigator::generatePathObjectives(void)
```

Comment:
Can this function be given data as args rather than using class members? Similar to one of Gareth's comments below, this makes it easier to follow the flow of data to know what might be used or mutated somewhere. It will also make this function testable.

Reply:
done